File: vidiaspot-marketplace/app/Services/NftService.php
<?php

namespace App\Services;

use App\Models\Nft;
use App\Models\NftCollection;
use App\Models\NftTransaction;
use App\Models\User;
use Illuminate\Support\Facades\Auth;

class NftService
{
    public function createCollection($collectionData, $userId = null)
    {
        $userId = $userId ?: Auth::id();
        
        return NftCollection::create([
            'name' => $collectionData['name'],
            'description' => $collectionData['description'] ?? '',
            'slug' => \Str::slug($collectionData['name']),
            'creator_id' => $userId,
            'owner_id' => $userId,
            'external_url' => $collectionData['external_url'] ?? null,
            'twitter_username' => $collectionData['twitter_username'] ?? null,
            'instagram_username' => $collectionData['instagram_username'] ?? null,
            'discord_url' => $collectionData['discord_url'] ?? null,
            'category' => $collectionData['category'] ?? 'art',
            'status' => 'active',
            'verified' => false,
            'blockchain' => $collectionData['blockchain'] ?? 'ethereum',
            'token_standard' => $collectionData['token_standard'] ?? 'ERC-721',
            'royalty_percentage' => $collectionData['royalty_percentage'] ?? 0,
            'metadata' => $collectionData['metadata'] ?? [],
        ]);
    }

    public function createNft($nftData, $collectionId = null, $userId = null)
    {
        $userId = $userId ?: Auth::id();
        
        $nft = Nft::create([
            'name' => $nftData['name'],
            'description' => $nftData['description'] ?? '',
            'collection_id' => $collectionId,
            'external_url' => $nftData['external_url'] ?? null,
            'image_url' => $nftData['image_url'] ?? null,
            'animation_url' => $nftData['animation_url'] ?? null,
            'token_id' => $nftData['token_id'] ?? $this->generateTokenId(),
            'contract_address' => $nftData['contract_address'] ?? null,
            'blockchain' => $nftData['blockchain'] ?? 'ethereum',
            'creator_id' => $userId,
            'owner_id' => $userId,
            'price' => $nftData['price'] ?? 0,
            'currency' => $nftData['currency'] ?? 'ETH',
            'status' => 'draft',
            'properties' => $nftData['properties'] ?? [],
            'levels' => $nftData['levels'] ?? [],
            'stats' => $nftData['stats'] ?? [],
            'is_listed' => $nftData['is_listed'] ?? false,
            'list_price' => $nftData['list_price'] ?? 0,
            'list_currency' => $nftData['list_currency'] ?? 'ETH',
            'royalty_percentage' => $nftData['royalty_percentage'] ?? 0,
            'royalty_recipient' => $nftData['royalty_recipient'] ?? $userId,
            'metadata' => $nftData['metadata'] ?? [],
            'token_standard' => $nftData['token_standard'] ?? 'ERC-721',
            'supply' => $nftData['supply'] ?? 1,
            'max_supply' => $nftData['max_supply'] ?? 1,
            'is_soulbound' => $nftData['is_soulbound'] ?? false,
            'transferable' => $nftData['transferable'] ?? true,
        ]);

        return $nft;
    }

    public function listNftForSale($nftId, $price, $currency = 'ETH', $userId = null)
    {
        $userId = $userId ?: Auth::id();
        
        $nft = Nft::where('id', $nftId)
            ->where('owner_id', $userId)
            ->first();

        if ($nft) {
            $nft->update([
                'is_listed' => true,
                'list_price' => $price,
                'list_currency' => $currency,
                'status' => 'active',
            ]);
        }

        return $nft;
    }

    public function buyNft($nftId, $buyerId = null)
    {
        $buyerId = $buyerId ?: Auth::id();
        
        $nft = Nft::where('id', $nftId)
            ->where('is_listed', true)
            ->where('status', 'active')
            ->first();

        if (!$nft || $nft->owner_id === $buyerId) {
            return null;
        }

        $sellerId = $nft->owner_id;
        $price = $nft->list_price;
        $currency = $nft->list_currency;

        // In a real implementation, this would handle the blockchain transaction
        // For now, we'll simulate the transaction
        
        \DB::beginTransaction();
        
        try {
            // Update NFT ownership
            $nft->update([
                'owner_id' => $buyerId,
                'is_listed' => false,
            ]);

            // Record transaction
            $transaction = NftTransaction::create([
                'nft_id' => $nftId,
                'from_user_id' => $sellerId,
                'to_user_id' => $buyerId,
                'transaction_type' => 'sale',
                'price' => $price,
                'currency' => $currency,
                'status' => 'success',
                'metadata' => [
                    'sale_price' => $price,
                    'previous_owner' => $sellerId,
                    'new_owner' => $buyerId,
                ],
            ]);

            \DB::commit();
            
            return ['nft' => $nft, 'transaction' => $transaction];
        } catch (\Exception $e) {
            \DB::rollback();
            throw $e;
        }
    }

    public function getMarketplaceNfts($filters = [])
    {
        $query = Nft::active()->listed()->with(['creator', 'owner', 'collection']);

        if (isset($filters['collection_id'])) {
            $query->where('collection_id', $filters['collection_id']);
        }

        if (isset($filters['blockchain'])) {
            $query->forBlockchain($filters['blockchain']);
        }

        if (isset($filters['creator_id'])) {
            $query->where('creator_id', $filters['creator_id']);
        }

        if (isset($filters['price_min'])) {
            $query->where('list_price', '>=', $filters['price_min']);
        }

        if (isset($filters['price_max'])) {
            $query->where('list_price', '<=', $filters['price_max']);
        }

        if (isset($filters['search'])) {
            $query->where('name', 'like', '%' . $filters['search'] . '%')
                  ->orWhere('description', 'like', '%' . $filters['search'] . '%');
        }

        return $query->orderBy('created_at', 'desc')
                    ->paginate($filters['per_page'] ?? 20);
    }

    public function getNftDetails($nftId)
    {
        return Nft::with(['creator', 'owner', 'collection', 'transactions'])->find($nftId);
    }

    public function getUserNfts($userId = null)
    {
        $userId = $userId ?: Auth::id();
        
        return Nft::where('owner_id', $userId)
            ->with(['collection'])
            ->orderBy('created_at', 'desc')
            ->get();
    }

    public function getUserCollections($userId = null)
    {
        $userId = $userId ?: Auth::id();
        
        return NftCollection::where('owner_id', $userId)
            ->orderBy('created_at', 'desc')
            ->get();
    }

    private function generateTokenId()
    {
        return (string) time() . rand(1000, 9999);
    }
}

File: vidiaspot-marketplace/mobile/lib/services/nft_service.dart
// lib/services/nft_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../models/nft/nft_model.dart';
import '../models/nft/nft_collection_model.dart';

class NftService {
  final String baseUrl = 'http://10.0.2.2:8000'; // For Android emulator, adjust as needed
  String? _authToken;

  NftService() {
    _loadAuthToken();
  }

  Future<void> _loadAuthToken() async {
    SharedPreferences prefs = await SharedPreferences.getInstance();
    _authToken = prefs.getString('auth_token');
  }

  Map<String, String> getHeaders() {
    Map<String, String> headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    if (_authToken != null) {
      headers['Authorization'] = 'Bearer $_authToken';
    }

    return headers;
  }

  // Get marketplace NFTs
  Future<List<Nft>> getMarketplaceNfts({
    String? collectionId,
    String? blockchain,
    int? creatorId,
    double? priceMin,
    double? priceMax,
    String? search,
    int perPage = 20,
  }) async {
    String url = '$baseUrl/nfts/marketplace?page=1&per_page=$perPage';
    
    if (collectionId != null) url += '&collection_id=$collectionId';
    if (blockchain != null) url += '&blockchain=$blockchain';
    if (creatorId != null) url += '&creator_id=$creatorId';
    if (priceMin != null) url += '&price_min=$priceMin';
    if (priceMax != null) url += '&price_max=$priceMax';
    if (search != null) url += '&search=$search';

    final response = await http.get(Uri.parse(url), headers: getHeaders());

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data['success'] ?? false) {
        return (data['nfts']['data'] as List)
            .map((json) => Nft.fromJson(json))
            .toList();
      } else {
        throw Exception(data['message'] ?? 'Failed to load NFTs');
      }
    } else {
      throw Exception('Failed to load NFTs: ${response.statusCode}');
    }
  }

  // Get NFT details
  Future<Nft> getNftDetails(int nftId) async {
    final response = await http.get(
      Uri.parse('$baseUrl/nfts/$nftId'),
      headers: getHeaders(),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return Nft.fromJson(data);
    } else {
      throw Exception('Failed to load NFT: ${response.statusCode}');
    }
  }

  // Get NFT collections
  Future<List<NftCollection>> getCollections({
    int? creatorId,
    String? category,
    bool? verified,
    int perPage = 20,
  }) async {
    String url = '$baseUrl/nft-collections?page=1&per_page=$perPage';
    
    if (creatorId != null) url += '&creator_id=$creatorId';
    if (category != null) url += '&category=$category';
    if (verified != null) url += '&verified=$verified';

    final response = await http.get(Uri.parse(url), headers: getHeaders());

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data['success'] ?? false) {
        return (data['collections']['data'] as List)
            .map((json) => NftCollection.fromJson(json))
            .toList();
      } else {
        throw Exception(data['message'] ?? 'Failed to load collections');
      }
    } else {
      throw Exception('Failed to load collections: ${response.statusCode}');
    }
  }

  // Get user's NFTs
  Future<List<Nft>> getUserNfts() async {
    final response = await http.get(
      Uri.parse('$baseUrl/nfts/user'),
      headers: getHeaders(),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data['success']) {
        return (data['nfts'] as List)
            .map((json) => Nft.fromJson(json))
            .toList();
      } else {
        throw Exception(data['message'] ?? 'Failed to load user NFTs');
      }
    } else {
      throw Exception('Failed to load user NFTs: ${response.statusCode}');
    }
  }

  // Get user's collections
  Future<List<NftCollection>> getUserCollections() async {
    final response = await http.get(
      Uri.parse('$baseUrl/nft-collections/user'),
      headers: getHeaders(),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data['success']) {
        return (data['collections'] as List)
            .map((json) => NftCollection.fromJson(json))
            .toList();
      } else {
        throw Exception(data['message'] ?? 'Failed to load user collections');
      }
    } else {
      throw Exception('Failed to load user collections: ${response.statusCode}');
    }
  }

  // Create a new NFT collection
  Future<NftCollection> createCollection({
    required String name,
    String? description,
    String? externalUrl,
    String? twitterUsername,
    String? instagramUsername,
    String? discordUrl,
    String category = 'art',
    String blockchain = 'ethereum',
    String tokenStandard = 'ERC-721',
    double royaltyPercentage = 0.0,
  }) async {
    final response = await http.post(
      Uri.parse('$baseUrl/nft-collections'),
      headers: getHeaders(),
      body: jsonEncode({
        'name': name,
        'description': description ?? '',
        'external_url': externalUrl,
        'twitter_username': twitterUsername,
        'instagram_username': instagramUsername,
        'discord_url': discordUrl,
        'category': category,
        'blockchain': blockchain,
        'token_standard': tokenStandard,
        'royalty_percentage': royaltyPercentage,
      }),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data['success']) {
        return NftCollection.fromJson(data['collection']);
      } else {
        throw Exception(data['message'] ?? 'Failed to create collection');
      }
    } else {
      throw Exception('Failed to create collection: ${response.statusCode}');
    }
  }

  // Create a new NFT
  Future<Nft> createNft({
    required String name,
    required String imageUrl,
    String? description,
    int? collectionId,
    String? externalUrl,
    String? animationUrl,
    String blockchain = 'ethereum',
    double price = 0.0,
    String currency = 'ETH',
    bool isListed = false,
    double listPrice = 0.0,
    String listCurrency = 'ETH',
    double royaltyPercentage = 0.0,
    Map<String, dynamic>? properties,
    Map<String, dynamic>? levels,
    Map<String, dynamic>? stats,
    int supply = 1,
    int maxSupply = 1,
    bool isSoulbound = false,
    bool transferable = true,
  }) async {
    final response = await http.post(
      Uri.parse('$baseUrl/nfts'),
      headers: getHeaders(),
      body: jsonEncode({
        'name': name,
        'description': description ?? '',
        'collection_id': collectionId,
        'external_url': externalUrl,
        'image_url': imageUrl,
        'animation_url': animationUrl,
        'blockchain': blockchain,
        'price': price,
        'currency': currency,
        'is_listed': isListed,
        'list_price': listPrice,
        'list_currency': listCurrency,
        'royalty_percentage': royaltyPercentage,
        'properties': properties ?? {},
        'levels': levels ?? {},
        'stats': stats ?? {},
        'supply': supply,
        'max_supply': maxSupply,
        'is_soulbound': isSoulbound,
        'transferable': transferable,
      }),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data['success']) {
        return Nft.fromJson(data['nft']);
      } else {
        throw Exception(data['message'] ?? 'Failed to create NFT');
      }
    } else {
      throw Exception('Failed to create NFT: ${response.statusCode}');
    }
  }

  // List NFT for sale
  Future<Nft> listNftForSale({
    required int nftId,
    required double price,
    String currency = 'ETH',
  }) async {
    final response = await http.post(
      Uri.parse('$baseUrl/nfts/$nftId/list'),
      headers: getHeaders(),
      body: jsonEncode({
        'price': price,
        'currency': currency,
      }),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data['success']) {
        return Nft.fromJson(data['nft']);
      } else {
        throw Exception(data['message'] ?? 'Failed to list NFT for sale');
      }
    } else {
      throw Exception('Failed to list NFT for sale: ${response.statusCode}');
    }
  }

  // Buy NFT
  Future<Map<String, dynamic>> buyNft(int nftId) async {
    final response = await http.post(
      Uri.parse('$baseUrl/nfts/$nftId/buy'),
      headers: getHeaders(),
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to buy NFT: ${response.statusCode}');
    }
  }

  // Get collection details
  Future<NftCollection> getCollectionDetails(int collectionId) async {
    final response = await http.get(
      Uri.parse('$baseUrl/nft-collections/$collectionId'),
      headers: getHeaders(),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return NftCollection.fromJson(data);
    } else {
      throw Exception('Failed to load collection: ${response.statusCode}');
    }
  }

  // Get NFTs in a collection
  Future<List<Nft>> getNftsInCollection(int collectionId) async {
    final response = await http.get(
      Uri.parse('$baseUrl/nft-collections/$collectionId/nfts'),
      headers: getHeaders(),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data['success']) {
        return (data['nfts'] as List)
            .map((json) => Nft.fromJson(json))
            .toList();
      } else {
        throw Exception(data['message'] ?? 'Failed to load collection NFTs');
      }
    } else {
      throw Exception('Failed to load collection NFTs: ${response.statusCode}');
    }
  }
}